{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04d3e2ce",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'fpdf'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mModuleNotFoundError\u001b[39m                       Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[4]\u001b[39m\u001b[32m, line 4\u001b[39m\n\u001b[32m      2\u001b[39m addpath = \u001b[33m\"\u001b[39m\u001b[33m/mnt/nas_1/YangLab/loci/tandem/stand_alone_consurf\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m      3\u001b[39m sys.path.append(addpath)\n\u001b[32m----> \u001b[39m\u001b[32m4\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mstand_alone_consurf\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m determine_mode\n\u001b[32m      6\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mget_form_direct\u001b[39m(\n\u001b[32m      7\u001b[39m     msa=\u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[32m      8\u001b[39m     query=\u001b[33m\"\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m   (...)\u001b[39m\u001b[32m     26\u001b[39m     cif=\u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[32m     27\u001b[39m ):\n\u001b[32m     28\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m work_dir \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/nas_1/YangLab/loci/tandem/stand_alone_consurf/stand_alone_consurf.py:13\u001b[39m\n\u001b[32m     11\u001b[39m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01msubprocess\u001b[39;00m\n\u001b[32m     12\u001b[39m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mtime\u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m13\u001b[39m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mfpdf\u001b[39;00m \n\u001b[32m     14\u001b[39m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mmath\u001b[39;00m\n\u001b[32m     15\u001b[39m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01margparse\u001b[39;00m\n",
      "\u001b[31mModuleNotFoundError\u001b[39m: No module named 'fpdf'"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "addpath = \"/home/newloci/tandem\"\n",
    "sys.path.append(addpath)\n",
    "from tandem.src.stand_alone_consurf.consurf import determine_mode, get_form_direct, extract_data_from_model, no_MSA, extract_data_from_MSA, find_best_substitution_model\n",
    "from tandem.src.stand_alone_consurf.consurf import run_rate4site_old, assign_colors_according_to_r4s_layers, write_MSA_percentage_file, consurf_create_output\n",
    "from tandem.src.stand_alone_consurf.consurf import conseq_create_output, zip_all_outputs\n",
    "from tandem.src.stand_alone_consurf.consurf import get_query_seq_in_MSA, reveal_buried_exposed\n",
    "import os, re\n",
    "import GENERAL_CONSTANTS\n",
    "from datetime import datetime\n",
    "from datetime import date\n",
    "import time\n",
    "from prody import LOGGER\n",
    "\n",
    "LOGGER.start(\"consurf.log\")\n",
    "\n",
    "form, vars = get_form_direct(\n",
    "    query=\"P69905\",\n",
    "    structure=\"4xr8.pdb\",\n",
    "    chain=\"A\",\n",
    "    DB=\"/home/newloci/alphafold3/db/uniref90_2022_05.fa\",\n",
    "    work_dir=\".\",\n",
    "    algorithm=\"HMMER\"\n",
    ")\n",
    "\n",
    "\n",
    "vars['run_log'] = \"log.txt\"\n",
    "# LOG = open(vars['working_dir'] + vars['run_log'], 'w')\n",
    "vars['tree_file'] = \"TheTree.txt\"\n",
    "vars['msa_fasta'] = \"msa_fasta.aln\" \n",
    "\n",
    "vars['running_mode'] = determine_mode(form, vars)\n",
    "vars['script_dir'] = os.path.abspath(\".\")\n",
    "\n",
    "os.chdir(vars['working_dir'])\n",
    "# LOG.write(\"determine_mode : Mode is: \" + vars['running_mode'] + \"\\n\")\n",
    "vars['BLAST_out_file'] = \"sequences_found.txt\"\n",
    "if form['DNA_AA'] == \"AA\": # proteins\n",
    "    vars['protein_or_nucleotide'] = \"proteins\"\n",
    "    vars['Msa_percentageFILE'] = \"msa_aa_variety_percentage.csv\"\n",
    "else: # nucleotides\n",
    "    vars['protein_or_nucleotide'] = \"nucleotides\"\n",
    "    vars['Msa_percentageFILE'] = \"msa_nucleic_acids_variety_percentage.csv\"\n",
    "\n",
    "\n",
    "vars['All_Outputs_Zip'] = \"Consurf_Outputs.zip\"\n",
    "if form['pdb_FILE'] is not None: # User PDB\n",
    "    match = re.search(r'([^\\/]+)$', form['pdb_FILE'])\n",
    "    if match:\n",
    "        vars['Used_PDB_Name'] = match.group(1)\n",
    "    vars['Used_PDB_Name'] = re.sub(r\"[() ]\", r\"_\", vars['Used_PDB_Name'])\n",
    "    match = re.search(r'(\\S+)\\.', vars['Used_PDB_Name'])\n",
    "    if match:\n",
    "        vars['Used_PDB_Name'] = match.group(1)\n",
    "else: # ConSeq Mode, No Model\n",
    "    vars['Used_PDB_Name'] = \"no_model\"\n",
    "\n",
    "\n",
    "if form['E_VALUE'].isdigit() and form['E_VALUE'] != \"0\":\n",
    "    # if the user inserted an integer, we turn it to a fraction\n",
    "    number_of_zeros = int(form['E_VALUE'])\n",
    "    form['E_VALUE'] = \"0.\"\n",
    "    i = 1\n",
    "    while i < number_of_zeros:\n",
    "        form['E_VALUE'] += \"0\"\n",
    "        i += 1\n",
    "    form['E_VALUE'] += \"1\"\n",
    "\n",
    "\n",
    "if int(form['MAX_REDUNDANCY']) >= 100:\n",
    "    vars['hit_redundancy'] = 99.999999\n",
    "else:\n",
    "    vars['hit_redundancy'] = float(form['MAX_REDUNDANCY'])\n",
    "\n",
    "\n",
    "form['Run_Number'] = \"0\"\n",
    "vars['hit_min_length'] = GENERAL_CONSTANTS.FRAGMENT_MINIMUM_LENGTH # minimum length of homologs\n",
    "vars['min_num_of_hits'] = GENERAL_CONSTANTS.MINIMUM_FRAGMENTS_FOR_MSA # minimum number of homologs\n",
    "vars['FINAL_sequences'] = \"query_final_homolougs.fasta\" # finial homologs for creating the MSA\n",
    "vars['FINAL_sequences_html'] = \"query_final_homolougs.html\" # html files showing the finial homologs to the user\n",
    "vars['submission_time'] = str(datetime.now())\n",
    "vars['date'] = date.today().strftime(\"%d/%m/%Y\")\n",
    "vars['time_table'] = []\n",
    "vars['current_time'] = time.time()\n",
    "vars['gradesPE'] = vars['Used_PDB_Name'] + \"_consurf_grades.txt\" # file with consurf output\n",
    "#vars['tree_file'] = \"TheTree.txt\"\n",
    "vars['zip_list'] = []\n",
    "\n",
    "# pymol and chimera scripts\n",
    "vars['chimera_color_script'] = vars['script_dir'] + \"color_consurf_chimerax_session.py\"\n",
    "vars['chimera_color_script_CBS'] = vars['script_dir'] + \"color_consurf_CBS_chimerax_session.py\"\n",
    "vars['pymol_color_script_isd'] = vars['script_dir'] + \"color_consurf_pymol_isd_session.py\"\n",
    "vars['pymol_color_script_CBS_isd'] = vars['script_dir'] + \"color_consurf_CBS_pymol_isd_session.py\"\n",
    "\n",
    "vars['msa_clustal'] = \"msa_clustal.aln\" # if the file is not in clustal format, we create a clustal copy of it\n",
    "\n",
    "vars['Colored_Seq_PDF'] = \"consurf_colored_seq.pdf\"\n",
    "vars['Colored_Seq_CBS_PDF'] = \"consurf_colored_seq_CBS.pdf\"\n",
    "\n",
    "vars['gradesPE_Output'] = [] # an array to hold all the information that should be printed to gradesPE\n",
    "# in each array's cell there is a hash for each line from r4s.res.\n",
    "# POS: position of that aa in the sequence ; SEQ : aa in one letter ;\n",
    "# GRADE : the given grade from r4s output ; COLOR : grade according to consurf's scale\n",
    "\n",
    "vars['zip_list'].append(vars['tree_file'])\n",
    "vars['zip_list'].append(vars['gradesPE'])\n",
    "vars['zip_list'].append(vars['Msa_percentageFILE'])\n",
    "vars['zip_list'].append(vars['Colored_Seq_PDF'])\n",
    "vars['zip_list'].append(vars['Colored_Seq_CBS_PDF'])\n",
    "vars['zip_list'].append(vars['msa_fasta'])\n",
    "\n",
    "## mode : include pdb\n",
    "\n",
    "# create a pdbParser, to get various info from the pdb file\n",
    "if vars['running_mode'] == \"_mode_pdb_no_msa\" or vars['running_mode'] == \"_mode_pdb_msa\" or vars['running_mode'] == \"_mode_pdb_msa_tree\":\n",
    "    extract_data_from_model(form, vars)\n",
    "\n",
    "\"\"\"\n",
    "## mode : only protein sequence\n",
    "\n",
    "# if there is only protein sequence: we upload it.\n",
    "elif vars['running_mode'] == \"_mode_no_pdb_no_msa\":\n",
    "\n",
    "    upload_protein_sequence()\n",
    "\"\"\"\n",
    "## mode : no msa - with PDB or without PDB\n",
    "\n",
    "if vars['running_mode'] == \"_mode_pdb_no_msa\" or vars['running_mode'] == \"_mode_no_pdb_no_msa\":\n",
    "    no_MSA(form, vars)\n",
    "## mode : include msa\n",
    "elif vars['running_mode'] == \"_mode_pdb_msa\" or vars['running_mode'] == \"_mode_msa\" or vars['running_mode'] == \"_mode_pdb_msa_tree\" or vars['running_mode'] == \"_mode_msa_tree\":\n",
    "    extract_data_from_MSA(form, vars)\n",
    "\n",
    "if form['SUB_MATRIX'] == \"BEST\":\n",
    "    #vars['best_fit'] = True\n",
    "    find_best_substitution_model(form, vars)\n",
    "else:\n",
    "    vars['best_fit'] = \"model_chosen\"\n",
    "\n",
    "run_rate4site_old(form, vars)\n",
    "assign_colors_according_to_r4s_layers(form, vars)\n",
    "write_MSA_percentage_file(form, vars)\n",
    "\n",
    "## mode : include pdb\n",
    "\n",
    "if vars['running_mode'] == \"_mode_pdb_no_msa\" or vars['running_mode'] == \"_mode_pdb_msa\" or vars['running_mode'] == \"_mode_pdb_msa_tree\":\n",
    "    consurf_create_output(form, vars)\n",
    "\n",
    "## mode : ConSeq - NO PDB\n",
    "\n",
    "if vars['running_mode'] == \"_mode_msa\" or vars['running_mode'] == \"_mode_no_pdb_no_msa\" or vars['running_mode'] == \"_mode_msa_tree\":\n",
    "    conseq_create_output(form, vars)\n",
    "\n",
    "zip_all_outputs(vars)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20d5099f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import sys\n",
    "# addpath = \"/home/newloci/tandem/src/stand_alone_consurf\"\n",
    "# sys.path.append(addpath)\n",
    "# from main import run\n",
    "from src.stand_alone_consurf.main import run \n",
    "consurfile = run(\n",
    "    query=\"P69905\",\n",
    "    structure=\"/home/newloci/tandem/4xr8.pdb\",\n",
    "    chain=\"A\",\n",
    "    DB=\"/home/newloci/alphafold3/db/uniref90_2022_05.fa\",\n",
    "    work_dir=\"P69905\",\n",
    "    algorithm=\"HMMER\"\n",
    ")\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
